### PR TITLE
WIP: Feature: creation of relocatable tarballs from installed packages

### PR DESCRIPTION
**Pull request by gartung**
This extends #445 which allows the creation of binary tarballs from installed packages and the installation of those tarballs on other machine with the appropriate rewriting of rpaths with patchelf on linux or install_name_tool on macos.

### CODE REVIEW

**Comment by gartung** [comment]
@hegner @tgamblin 
I have merged from llnl/develop and fixed any merge problems. This incorporates #445. I think this is ready for merge to develop.

File: lib/spack/spack/binary_distribution.py [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+    Download binary tarball for given package into stage area
+    Return True if successful
+    """
+    if len(spack.config.get_config('mirrors')) == 0:
+        tty.die("Please add a spack mirror to allow " +
+                "download of pre-compiled packages.")
+
+    # stage the tarball into standard place
+    tarball = tarball_path_name(package.spec)
+    stage = Stage(tarball, name=package.stage.path, mirror_path=tarball)
```
Comment (gartung) [inline lib/spack/spack/binary_distribution.py]: @tgamblin 
This is now working anymore. It like Stage is trying to interpret the following example as a URL

macos1011-x86_64/clang-7.3.0-apple/libsigsegv/macos1011-x86_64-libsigsegv-2.10-ftmuuyrg2q6e2hjaokozwhrv46e5dq42.tar.gz

The error message is
==> Error: clang-7.3.0-apple does not match any installed packages.

How can I proceed. The idea is to pull the binary tarball from a mirror.

File: lib/spack/spack/binary_distribution.py [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+    try:
+        stage.fetch(mirror_only=True)
```
Comment (gartung) [inline lib/spack/spack/binary_distribution.py]: @tgamblin I was incorrect. The error is happening in this call.

**Comment by gartung** [comment]
@tgamblin  I found a resolution to the problem of downloading the binary tarball. I loop over the mirror urls and add that to the tarball path and then pass that to Stage. I also merged llnl/develop recently and resolved any merge conflicts.

File: lib/spack/spack/architecture.py [hunk 3]
```diff
@@ -487,6 +487,27 @@ def arch_from_dict(d):
     return arch
 
 
+def get_full_system_from_platform():
```
Comment (tgamblin) [inline lib/spack/spack/architecture.py]: It looks like the tarballs use this for the name, but Spack now has `spack.platform`, and `spack.architecture.sys_type()` will give you a platform-os-target string like `darwin-elcapitan-x86_64`.  On cray the target can be more specific (`haswell`, `interlagos`, etc.) and we're planning to allow that level of specificity on other platforms (Linux, Mac) too.

Is that sufficient or do you guys need a separate platform descriptor for the binaries?

Comment (hegner) [inline lib/spack/spack/architecture.py]: @tgamblin - if that contains os name, version, and the processor arch that's sufficient, yes.

File: lib/spack/spack/binary_distribution.py [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+
+def prepare():
+    """
+    Install patchelf as pre-requisite to the
+    required relocation of binary packages
+    """
+    if platform.system() == 'Darwin':
+        return
+    dir = os.getcwd()
+    patchelf_spec = spack.cmd.parse_specs("patchelf", concretize=True)[0]
```
Comment (tgamblin) [inline lib/spack/spack/binary_distribution.py]: `Spec('patchelf').concretized()` might be simpler here, but there's a better way now

Comment (hegner) [inline lib/spack/spack/binary_distribution.py]: @tgamblin - thanks, didn't know. What do you mean with "there's a better way now"?

File: lib/spack/spack/binary_distribution.py [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+    if not spack.install_layout.check_installed(patchelf_spec):
```
Comment (tgamblin) [inline lib/spack/spack/binary_distribution.py]: `spack.installed_db` has a `query()` method.  You can just run this:

```python
installed_specs = spack.installed_db.query('patchelf')
```

You probably really want to run:
```python
installed_specs = spack.installed_db.query('patchelf os=frontend')
```

to ensure you get a patchelf that matches the default frontend OS for the system (e.g. SuSE and not CNL on Cray).

File: lib/spack/spack/binary_distribution.py (hunk 4 above)
Comment (tgamblin) [inline lib/spack/spack/binary_distribution.py]: Adding comments as I go 😄 -- I mean you don't need to concretize this.  `installed_db.query()`, described below, should match against specs that are installed, even if they don't exactly match what `patchelf` currently concretizes to.

File: lib/spack/spack/relocate.py [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+    else:
+        tty.die("Relocation not implemented for %s" % platform.system())
+
+
+def relocate_text(path_name, old_dir, new_dir):
+    """
+    Replace old path with new path in text files
+    """
+    os.system("LC_ALL=C sed -i -e \"s#%s#%s#g\" \"%s\""
+              % (old_dir, new_dir, path_name))
```
Comment (tgamblin) [inline lib/spack/spack/relocate.py]: This should probably use `filter_file` (which spack packages use) from `llnl.util.filesystem`.

File: var/spack/repos/builtin.mock/packages/patchelf/package.py [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+# conditions of the GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with this program; if not, write to the Free Software
+# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+##############################################################################
+from spack import *
+
+
+class Patchelf(Package):
```
Comment (tgamblin) [inline var/spack/repos/builtin.mock/packages/patchelf/package.py]: This has since been added to Spack (sorry for the slow PR review!)

File: lib/spack/spack/package.py [hunk 8]
```diff
@@ -938,8 +962,11 @@ def build_process():
                 # Run the pre-install hook in the child process after
                 # the directory is created.
                 spack.hooks.pre_install(self)
-
-                if fake:
+                # Set up process's build environment before running install.
```
Comment (tgamblin) [inline lib/spack/spack/package.py]: Should probably also emit a different message from `tty.msg("Building %s" % self.name)` on line 954 above.  User should be told that a binary cache is being installed instead.

File: lib/spack/spack/cmd/install.py [hunk 9]
```diff
@@ -61,7 +62,11 @@ def setup_parser(subparser):
     subparser.add_argument(
         '--run-tests', action='store_true', dest='run_tests',
         help="Run tests during installation of a package.")
-
+    subparser.add_argument(
+        '-p', '--install-policy', action='store', dest="install_policy",
```
Comment (tgamblin) [inline lib/spack/spack/cmd/install.py]: docs still say `--policy` but this says `--install-policy`.  `--policy` seems too general, but `--install-policy` seems redundant when the command is `spack install`.  What about `--binary=[never|lazy|always]`?

File: lib/spack/spack/binary_distribution.py [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+
+
+def write_buildinfo_file(spec):
+    """
+    Create a cache file containing information
+    required for the relocation
+    """
+    text_to_relocate   = []
+    binary_to_relocate = []
+    blacklist = (".spack", "man")
```
Comment (tgamblin) [inline lib/spack/spack/binary_distribution.py]: What if they put binaries in `man`?  I mean, this is HPC... I've seen a lot of crazy things.

File: lib/spack/spack/binary_distribution.py [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+                         str(spec.compiler).replace("@", "-"),
+                         spec.name)
+
+
+def tarball_name(spec):
+    """
+    Return the name of the tarfile according to the convention
+    <os>-<architecture>-<package>-<dag_hash>.tar.gz
+    """
+    return "%s-%s-%s-%s.tar.gz" % (get_full_system_from_platform(),
```
Comment (tgamblin) [inline lib/spack/spack/binary_distribution.py]: This should use `spack.architecture.sys_type()`, which contains the "platform", OS, and target... Any reason the package isn't first?  What about:

`<package>-<platform>-<os>-<target>-<dag_hash>.tar.gz`

File: lib/spack/spack/binary_distribution.py [hunk 12]
```diff
@@ -0,0 +1,209 @@
...
+        mkdirp(tarfile_dir)
+
+    tar = which('tar', required=True)
+    dirname = os.path.dirname(spec.prefix)
+    basename = os.path.basename(spec.prefix)
+
+    # handle meta-data
+    cp = which("cp", required=True)
+    spec_file = os.path.join(spec.prefix, ".spack", "spec.yaml")
+    target_spec_file = tarfile + ".yaml"
```
Comment (tgamblin) [inline lib/spack/spack/binary_distribution.py]: is this for easily querying the specs once the tarballs are generated?  I couldn't see where this gets used.

File: lib/spack/spack/binary_distribution.py [hunk 13]
```diff
@@ -0,0 +1,209 @@
...
+
+def write_buildinfo_file(spec):
+    """
+    Create a cache file containing information
+    required for the relocation
+    """
+    text_to_relocate   = []
+    binary_to_relocate = []
+    blacklist = (".spack", "man")
+    for root, dirs, files in os.walk(spec.prefix, topdown=True):
```
Comment (tgamblin) [inline lib/spack/spack/binary_distribution.py]: If it's not ridiculously expensive, we should consider just making this file get generated and installed to `.spack` after every install, so it's easy to make a tarball later.

File: lib/spack/spack/binary_distribution.py (hunk 10 above)
Comment (hegner) [inline lib/spack/spack/binary_distribution.py]: We removed it because it was just time wasted in there. Scanning plenty of files. I will remove that exclusion.

File: lib/spack/spack/binary_distribution.py (hunk 13 above)
Comment (hegner) [inline lib/spack/spack/binary_distribution.py]: I'd prefer doing that once we went through the first iterations and know the format is really stable.

File: lib/spack/spack/binary_distribution.py (hunk 11 above)
Comment (hegner) [inline lib/spack/spack/binary_distribution.py]: It is for convenience as this is how we order/group the binary archives in the end. From single package perspective your comment makes sense though.

File: lib/spack/spack/binary_distribution.py [hunk 14]
```diff
@@ -0,0 +1,209 @@
...
+    mirrors=spack.config.get_config('mirrors')
+    if len(mirrors) == 0:
+        tty.die("Please add a spack mirror to allow " +
+                "download of pre-compiled packages.")
+    tarball = tarball_path_name(package.spec)
+    for key in mirrors:
+        url=mirrors[key]+"/"+tarball
+        print url
+        # stage the tarball into standard place
+        stage = Stage(url, name=package.stage.path)
```
Comment (tgamblin) [inline lib/spack/spack/binary_distribution.py]: This will construct a `URLFetchStrategy` with no associated checksum, and it'll just trust the remote source that the binary is ok.  Before we can put this in the mainline, we need to figure out a way to embed checksums in the packages, although that could result in a lot of information as the number of binary builds grows.  

Homebrew's bottle scheme is pretty simple, and it makes it pretty easy to put binary hashes into formulae.  Spack has a slightly more difficult situation because a) there can be a lot of different builds for the same package and b) you can't get the symbolic `spec` straight from a hash.

I think a) and b) can eventually be mitigated if binary archive sites had a way to provide indices of available hashes, so that Spack could download an index of available binaries and find a "close enough match".  For a first cut at this, exact matching of the DAG hash is probably fine, especially if you expect users typically not to pull straight from `develop` (which would mean that things don't change as much).

File: lib/spack/spack/binary_distribution.py (hunk 12 above)
Comment (hegner) [inline lib/spack/spack/binary_distribution.py]: Yap. This is there to facilitate the mirroring and querying. You don't want to extract the data from every package "manually" again.

File: lib/spack/spack/cmd/install.py (hunk 9 above)
Comment (hegner) [inline lib/spack/spack/cmd/install.py]: Sounds better. What about `--cached-binary` as `binary` again is a little bit too generic?

File: lib/spack/spack/binary_distribution.py (hunk 13 above)
Comment (tgamblin) [inline lib/spack/spack/binary_distribution.py]: sounds reasonable

File: lib/spack/spack/relocate.py (hunk 6 above)
Comment (hegner) [inline lib/spack/spack/relocate.py]: Thanks. :-)

File: lib/spack/spack/binary_distribution.py (hunk 10 above)
Comment (tgamblin) [inline lib/spack/spack/binary_distribution.py]: If it saves tons of time it might be worth it 😄.  I actually _haven't_ seen anyone put binaries in `man`, but it does seem like a good idea to be complete.

File: lib/spack/spack/binary_distribution.py (hunk 12 above)
Comment (tgamblin) [inline lib/spack/spack/binary_distribution.py]: Ok this would be useful, then, for partial matching if we implement that.  This can be used to check if a spec is "good enough" to satisfy a build.

File: lib/spack/spack/cmd/install.py (hunk 9 above)
Comment (tgamblin) [inline lib/spack/spack/cmd/install.py]: `--fetch-binary`?

Comment (hegner) [inline lib/spack/spack/cmd/install.py]: perfect.

File: lib/spack/spack/binary_distribution.py (hunk 14 above)
Comment (hegner) [inline lib/spack/spack/binary_distribution.py]: @tgamblin - seems you are mixing quite a few concepts into your comment and I am not sure I get all the points you are trying to make. I'd believe that is something for the telcon. 
And yes, mirrors should have a way to fetch trustworthy hashes from there. I don't think the homebrew approach will work. The extraction of the metadata yaml was a first step into that direction.

**Comment by tgamblin** [comment]
Hi @gartung @hegner: This PR looks awesome.  I think most of what is needed is in here -- see comments above.

The one thing that's missing is how we'll make this secure.  I think the binary fetching needs to be checksummed or checked by some other means before we turn this loose on users -- see my [last comment](https://github.com/LLNL/spack/pull/1013#discussion_r74391075).

I am not sure what the best strategy is here.  As I mentioned, checksums might get unwieldy.  Another option would be to actually sign the packages.  [Debian uses GPG signatures](http://blog.packagecloud.io/eng/2014/10/28/howto-gpg-sign-verify-deb-packages-apt-repositories/) for this.  I worry about signatures because they can get clunky, but it might be worthwhile vs. maintaining a large number of checksums in each `package.py`.  I think the signature scheme might also integrate better with the plans @hegner and I were discussing to query a mirror for available packages first, then select a suitable one from what's available. Debian's signing of repository metadata looks similar to how I think we'd implement that.

What do you think?

File: lib/spack/spack/binary_distribution.py (hunk 14 above)
Comment (tgamblin) [inline lib/spack/spack/binary_distribution.py]: Ok -- we can discuss more on the call.  I think it's worthwhile to have a more detailed discussion about this.

I'm not sure you want to actually fetch the hashes directly from the mirror.  You need to have a way to guarantee that the hash itself came from a secure channel.  If you trust the server to provide both a reliable checksum and a reliable package, you're vulnerable to a compromised server.  The hashes/other verification mechanism need to come along some secure channel.  The approach I mentioned of signing packages like Debian seems promising for this as it's easy to transfer a public key once and verify packages with it from then on.  

btw -- LLNL has some S3 space where we can start to put up a mirror.  CERN may have something similar -- we can talk about that on the call as well.

Comment (hegner) [inline lib/spack/spack/binary_distribution.py]: Correct - trustworthy is the point. More for the telcon.

What concerns mirrors - we could set up something for a start.

**Comment by hegner** [comment]
Thanks. I think all you say is reasonable. Let's discuss the strategy and then we can implement it.

File: lib/spack/spack/binary_distribution.py (hunk 14 above)
Comment (gartung) [inline lib/spack/spack/binary_distribution.py]: @tgamblin I added this kludge because the previous method of fetching the tarball from a mirror was broken after merging in llnl/develop. I have been using my linux desktop as the mirror for testing but it only works onsite at Fermilab. The hashes are there so we can match the binary built to the spec used to build it.

**Comment by gartung** [comment]
I took a look at the distribution servers that are used for RPM's and binary tarball's for the two software systems I use regularly. Our only verification is that the install scripts and files come from known servers.

**Comment by hartzell** [comment]
**I'd love to see this get merged!**

I played around with a bit and it seems to work as advertised.

The only problem that I saw (which might well be PEBKAC or unrelated) was an inability to remove a broken mirror that I added:

```
[throgg@spack-central ~]$ spack mirror list
testing    file:///home/throgg/foo/tmp/caches
[throgg@spack-central ~]$ spack mirror remove testing
==> Removed mirror testing with url file:///home/throgg/foo/tmp/caches
[throgg@spack-central ~]$ spack mirror list
testing    file:///home/throgg/foo/tmp/caches
```

I worked around it with `vi`.

I'm not urgently worried about checksums/signatures and would be happy to see that come in down the line.

**Comment by gartung** [comment]
Try ‘spack config edit mirrors’ and remove the testing line. Oh wait, you did that with vi.

**Comment by hartzell** [comment]
> Try ‘spack config edit mirrors’ and remove the testing line. Oh wait, you did that with vi.

Yep.  `vi`.  Didn't seem worth busting out `emacs`.

`spack mirror remote testing` should have worked though, right?  Or at least not appeared to have suceeded?

g.

**Comment by gartung** [comment]
It should, but doesn’t.

**Comment by hartzell** [comment]
*phew*

**Comment by citibeth** [comment]
This looks like neat stuff, and clearly there's a lot of desire to see it merged.  The docs as written need to be reorganized / placed in context within the larger Spack.  @tgamblin should I grab the docs from this PR and integrate them now on the assumption this will be merged soon?  Or should we hold off...?

**Comment by tgamblin** [comment]
I plan to get this merged and also think it's awesome, but let's not try to integrate it with the current doc revisions

**Comment by gartung** [comment]
@tgamblin What do you think of putting this code into a stand-alone python file that can be called directly, like spack-dist.py?

**Comment by gartung** [comment]
@tgamblin  "spack install -fb always foo" only pulls the tarball for the deepest/first dependency in the tree. Everything else in built event with -fb always.

**Comment by gartung** [comment]
I give up!

**Comment by adamjstewart** [comment]
I think the reason the documentation tests are failing is because you don't link to `binary_caches.rst` in `index.rst`. So it won't be displayed on the website unless you manually type the URL.

**Comment by gartung** [comment]
Once more into the breach!

**Comment by alalazo** [comment]
@gartung Is this being reworked somewhere else? Or are you closing this PR for good?

**Comment by gartung** [comment]
It will probably need reworking once signing is merge in.

**Comment by tgamblin** [comment]
@mathstuf is going to work GPG signing into this and we'll get it merged.

**Comment by mathstuf** [comment]
My thought for making the packages signed is the following:

`<name>.spack` (a regular `.tar` file) containing the `tar.gz` (or `.bz2`, `.xz`, etc.), a `spec.yaml`, and each of those has a corresponding `.asc` file containing the signature of each. Then `spack create-tarball` would also take an optional `--key` option to sign with a specific key. Before installation, the spec and binary would be verified against the trusted key in the user's Spack keyring.

Thoughts?

**Comment by citibeth** [comment]
> <name>.spack (a regular .tar file) containing the tar.gz (or .bz2, .xz,
> etc.), a spec.yaml, and each of those has a corresponding .asc file
> containing the signature of each. Then spack create-tarball would also
> take an optional --key option to sign with a specific key. Before
> installation, the spec and binary would be verified against the trusted key
> in the user's Spack keyring.
>
> Thoughts?
>

You probably want to include all relevant packages files (package.py,
patch, etc) along with spec.yaml.  That make it truly portable between
different versions of Spack.

**Comment by gartung** [comment]
Seems reasonable to me. This would allow the person creating the tarball to sign using their own private key and put the public key on the site hosting the tarballs.

**Comment by gartung** [comment]
The .spack subdirectory of an installed package has a copy of the package.py for the package itself and any direct dependencies. This is included in the tarball.

**Comment by citibeth** [comment]
Awesome, thanks!

**Comment by gartung** [comment]
This may be an ignorant question... How do you sign using only the default python functions? It looks like PGP is an extra python package.

**Comment by gartung** [comment]
I guess you would have to add PGP as a package and install it as part of the tarball prep.

**Comment by mathstuf** [comment]
See #3845. There's a TODO item in there for using a spack-built `gpg2` package if `PATH` doesn't already contain it.

**Comment by gartung** [comment]
Cool thanks. I was about to add a GnuPG packages to spack using
https://gnupg.org/ftp/gcrypt/gnupg/gnupg-2.1.21.tar.bz2

**Comment by mathstuf** [comment]
Once #3845 is merged, this can be updated to have it and then we can work on getting the signing steps implemented.

File: lib/spack/spack/binary_distribution.py [hunk 15]
```diff
@@ -0,0 +1,230 @@
...
+    if os.path.exists(tarfile):
+        if force:
+            os.remove(tarfile)
+        else:
+            tty.warn("file exists, use -f to force overwrite: %s" % tarfile)
+            return
+    if not os.path.exists(tarfile_dir):
+        mkdirp(tarfile_dir)
+
+    tar = which('tar', required=True)
```
Comment (mathstuf) [inline lib/spack/spack/binary_distribution.py]: Is there a reason we're using the command line and not the `tarfile` module?

Comment (gartung) [inline lib/spack/spack/binary_distribution.py]: No. It may not have been available at the inception of this branch.

Comment (adamjstewart) [inline lib/spack/spack/binary_distribution.py]: I believe @mathstuf is referring to https://docs.python.org/2/library/tarfile.html

Comment (adamjstewart) [inline lib/spack/spack/binary_distribution.py]: For the record, we use tar from the command line in other places across Spack. Are there any advantages to using `tarfile`?

Comment (gartung) [inline lib/spack/spack/binary_distribution.py]: I did not author that part so I was guessing.

Comment (mathstuf) [inline lib/spack/spack/binary_distribution.py]: Using the module, you can add contents to the tarball without physically moving them to a single place on disk.

Comment (mathstuf) [inline lib/spack/spack/binary_distribution.py]: It also means you don't accidentally rely on GNU tar command line features.

Comment (gartung) [inline lib/spack/spack/binary_distribution.py]: True. On macOS command line options can be different than their Linux counterparts.

Comment (mathstuf) [inline lib/spack/spack/binary_distribution.py]: Please see mathstuf/spack@655ab426606f4c80d1ec9f3e92bf446f47d1c478 for a combination of the `tarfile` usage and GPG signing hookin points.

Comment (gartung) [inline lib/spack/spack/binary_distribution.py]: OK, you stripped out some things that are needed. I will work with what you created so far.

Comment (gartung) [inline lib/spack/spack/binary_distribution.py]: It still expects the binary_distribution file

==> Error: [Errno 2] No such file or directory: '/home/gartung/spack/opt/spack/linux-rhel7-x86_64/gcc-4.8.5/gcc-6.3.0-q2xyahg63sythq4blzln53n3fifr6qco/.spack/binary_distribution'

Comment (mathstuf) [inline lib/spack/spack/binary_distribution.py]: Hmm, does the call to `write_buildinfo_file` not handle this? Or does that write to somewhere not already under the prefix of the install we're bundling up?

Comment (gartung) [inline lib/spack/spack/binary_distribution.py]: Debugging now.

Comment (gartung) [inline lib/spack/spack/binary_distribution.py]: If the <prefix>/.speck/binary_distribution file exists this fails. Renaming/removing the binary_distribution file move me to the next error 

==> Error: [Errno 2] No such file or directory: 'gcc-6.3.0-jtpkgvgufggz574wytgnxtzcqgx4zvvw'

Comment (gartung) [inline lib/spack/spack/binary_distribution.py]: Need to default tarfile_path to current working directory.

Comment (gartung) [inline lib/spack/spack/binary_distribution.py]: Actually what appears to be missing is the path to the tarfile being created.

Comment (gartung) [inline lib/spack/spack/binary_distribution.py]: Bug is in create_tarball.py. Need to set the default output directory to $CWD there.

Comment (gartung) [inline lib/spack/spack/binary_distribution.py]: Or it is elsewhere. Tracing now.

Comment (gartung) [inline lib/spack/spack/binary_distribution.py]: @mathstuf python tarfile is much slower than command line tar.

Comment (gartung) [inline lib/spack/spack/binary_distribution.py]: Actually tarfile is fine. I was creating a tarball of everything installed.

File: lib/spack/spack/binary_distribution.py [hunk 16]
```diff
@@ -0,0 +1,260 @@
...
+    with tarfile.open(tarfile_path, 'w:gz') as tar:
+        tar.add(name = '%s' % spec.prefix 
+               , arcname = '%s' % os.path.basename(spec.prefix))
+
+    # Sign the packages.
+    #spack gpg sign [--key key] tarfile_path
+    #spack gpg sign [--key key] tarfile_path + '/spec.yaml'
+    path1='%s.asc' % tarfile_path
+    with open(path1,'a'):
+        os.utime(path1, None)
```
Comment (mathstuf) [inline lib/spack/spack/binary_distribution.py]: Why is this required? Or is this just in lieu of `spack gpg` not being available yet?

File: lib/spack/spack/binary_distribution.py [hunk 17]
```diff
@@ -0,0 +1,260 @@
...
+    if old_path == new_path:
+        return True  # No need to relocate
+
+    tty.warn("Using experimental feature for relocating package from",
+             "%s to %s." % (old_path, new_path))
+
+    # as we may need patchelf, find out where it is
+    patchelf_executable = ''
+    if platform.system() != 'Darwin':
+        patchelf_spec = spack.cmd.parse_specs("patchelf", concretize=True)[0]
```
Comment (mathstuf) [inline lib/spack/spack/binary_distribution.py]: This is cool; I'll port it over to the GPG branch.

File: lib/spack/spack/binary_distribution.py (hunk 16 above)
Comment (gartung) [inline lib/spack/spack/binary_distribution.py]: This is a mockup to test adding and extracting the .asc files.

File: lib/spack/spack/binary_distribution.py (hunk 17 above)
Comment (gartung) [inline lib/spack/spack/binary_distribution.py]: You can thank @hegner

Comment (mathstuf) [inline lib/spack/spack/binary_distribution.py]: So it actually isn't working here; it keeps throwing `NoCompilersForArchError: No compilers found for operating system debian6 and target x86_64.` on me. I have `fedora25-x86_64`. How do I get it to just use any compiler?

Comment (hegner) [inline lib/spack/spack/binary_distribution.py]: I'm a little bit behind in following the developments here. It fails in the concretization of that spec. Any compiler would be fine if there is one for your platform. Question would be why you get debian6 on one side and fedora25 on another. Does a command line install of patchelf work?

Comment (mathstuf) [inline lib/spack/spack/binary_distribution.py]: It also seems that the cmd infrastructure just swallows exceptions? I see the exceptions in the test, but the command just silently does nothing :/ . `tty` also doesn't actually output anything.

Comment (mathstuf) [inline lib/spack/spack/binary_distribution.py]: It isn't patchelf, but `gnupg@2` I'm doing (same results without the `@2`). But yes, a command line install seems to work.

Comment (hegner) [inline lib/spack/spack/binary_distribution.py]: Since you have it now installed, can you check whether `spack.installed_db.query("gnupg")` inside the code yields the proper result?

Comment (mathstuf) [inline lib/spack/spack/binary_distribution.py]: This returns the right thing: `spack.database.Database('opt/spack').query('gnupg@2:')`, but I cannot call `concretize` on a spec :/ .

**Comment by mathstuf** [comment]
Might it be better to have a command structure like `spack gpg` like `spack package create`, `spack package upload`, and `spack package install`?

**Comment by gartung** [comment]
Let me take a look at see if I can figure out how you did that.

**Comment by gartung** [comment]
How do you verify extra positional arguments exist?

**Comment by mathstuf** [comment]
I'm not sure what you mean? Do you have an example of what you expect to happen?

Preemptively: How are you doing so now? Why would that not be applicable to the subcommand parsers as well?

**Comment by gartung** [comment]
I figured out what I was doing wrong. Thanks,

**Comment by gartung** [comment]
I will be adding the ability to find what build caches are available on the mirror and install based on the spec.yaml included in the build cache.

**Comment by gartung** [comment]
I added the command "spack buildcache list" which will list all of the specs for the current machines architecture and a command that can be copied and pasted to do the actual buildcache install. I am also think about ways that the install path can be passed to "spack buildcache install" and possibly an alternate rpath prefix. I think this would all the installation on cvmfs where the server path and the client path are different.

**Comment by gartung** [comment]
I added the ability to change the absolute rpaths to relative rpaths with 
spack buildcache create --rel <spec>
The spec can be the has of an installed package. 
spack buildcache list 
will search all of the configured mirrors and grab the yaml files for all of the build caches available and prints a list of commands you can use to install the build cache.
spack buildcache install <spec>
will install the build cache and its dependent build caches using the hash of the spec.

**Comment by gartung** [comment]
@mathstuf I couldn't figure out how to sign the .spack or .yaml files.

**Comment by gartung** [comment]
@mathstuf I added the gpg2 signing and verification as options.

**Comment by gartung** [comment]
@mathstuf 
Should I include this file?
spack/opt/spack/gpg/pubring.kbx

**Comment by mathstuf** [comment]
Hmm, there should probably be a `.gitignore` for that file. I'll make a PR.

As for signing, I think it shouldn't be an option. Packages should always be signed when created and there should be a (hidden?) option to skip verification of packages when installing them.

**Comment by mathstuf** [comment]
Oh, that file is under `opt`. That entire directory should be ignored already.

**Comment by gartung** [comment]
I see that there is a keys directory in
spack/var/spack/gpg.mock/keys/external.key

Should I put my public key in 
spack/var/spack/gpg/keys

**Comment by gartung** [comment]
Does the buildcache command need an upload subcommand? We usually do that by hand or with and external script.

**Comment by gartung** [comment]
@mathstuf 
Can spack/opt/spack/gpg/ be moved to another spack install? I want to preserve it to I don't have to keep adding public keys for every spack install where I have run "spack pgp create".

**Comment by mathstuf** [comment]
> Can spack/opt/spack/gpg/ be moved to another spack install?

Yes, that directory should be relocatable (it's just a spack-specific `$HOME/.gnupg` directory).

File: lib/spack/spack/binary_distribution.py [hunk 18]
```diff
@@ -337,3 +337,26 @@ def get_specs():
...
+        p, links = spider(url + "/build_cache")
+        for link in links:
+            if re.search("\.key", link):
+                with Stage(link, name="build_cache", keep=True) as stage:
+                    try:
+                        stage.fetch()
+                    except fs.FetchError:
+                        next
+                if install:
+                    Gpg.trust(stage.save_filename)
```
Comment (mathstuf) [inline lib/spack/spack/binary_distribution.py]: I think there should be a verification before this. Something like:

```
This package is signed by this key:

<key info>

Do you want to trust this key? [y/N] > 
```

Comment (mathstuf) [inline lib/spack/spack/binary_distribution.py]: Or `this key was found on mirror <mirror name>`.

**Comment by becker33** [comment]
@gartung it looks to me like codecov is reporting coverage properly. I find the diff view in codecov is the easiest to interpret.

I would certainly like to see more tests for this feature, with the caveat that I have not yet reviewed it sufficiently to have any understanding of how difficult it will be to create those tests.

**Comment by adamjstewart** [comment]
Another thing to clarify about the Codecov values is that they are the comparison of this branch with the current develop, not with develop at the time this PR branched off. If coverage has increased in develop since you branched off, it will show up in the diff here. So if you rebase on the tip of develop, you'll get a better idea at how the global coverage is being affected. With that said, the `codecov/patch` is probably accurate.

**Comment by gartung** [comment]
So the red in the diff view is code that is not covered by a test?

**Comment by adamjstewart** [comment]
Well, the red in the "Files changed" view is lines that were removed, but if you install the [Codecov](https://chrome.google.com/webstore/detail/codecov-extension/keefkhehidemnokodkdkejapdgfjmijf?hl=en-US) extension, the red lines will be lines not covered by unit tests.

**Comment by becker33** [comment]
The muted red/green background in the "diff" view is the changes to the code, but the bright red/green bar on the left is covered/not covered lines of code

**Comment by gartung** [comment]
@adamjstewart I have the extension installed now in chrome, but the github page for my pull request shows "not covered" in the menu bar. How do I enable it?

**Comment by gartung** [comment]
I am going to resubmit once I get this code coverage thing figured out.
